### PR TITLE
fix: add support for react native 0.71

### PR DESCRIPTION
**Pull request by satya164**

### CODE REVIEW

File: packages/create-react-native-library/templates/native-common/android/build.gradle [hunk 1]
```diff
@@ -183,11 +178,17 @@ def kotlin_version = getExtOrDefault('kotlinVersion')
 <% } -%>
 
 dependencies {
+  if (REACT_NATIVE_MAJOR_VERSION == 0 && REACT_NATIVE_MINOR_VERSION < 71) {
     //noinspection GradleDynamicVersion
-  implementation "com.facebook.react:react-native:+"
+    implementation 'com.facebook.react:react-native:+'
+    // From node_modules
+  } else {
+    implementation 'com.facebook.react:react-android:' + REACT_NATIVE_VERSION
```
Comment (cortinico) [inline packages/create-react-native-library/templates/native-common/android/build.gradle:186]:
```suggestion
    implementation 'com.facebook.react:react-android'
```

File: packages/create-react-native-library/templates/native-common/android/build.gradle [hunk 2]
```diff
@@ -112,69 +152,24 @@ repositories {
...
-      parentDir = parentDir.parentFile
-
-      def androidSourcesDir = new File(
-        parentDir,
-        'node_modules/react-native'
-      )
-
-      def androidPrebuiltBinaryDir = new File(
-        parentDir,
-        'node_modules/react-native/android'
```
Comment (cortinico) [inline packages/create-react-native-library/templates/native-common/android/build.gradle:150]: I'm looking into re-creating the `react-native/android/` folder just to short-circut this logic.

But yes it should be cleaned up.

File: packages/create-react-native-library/templates/native-common/android/build.gradle [hunk 3]
```diff
@@ -112,69 +152,24 @@ repositories {
...
-  }
+  if (REACT_NATIVE_MAJOR_VERSION == 0 && REACT_NATIVE_MINOR_VERSION < 71) {
+    def androidSourcesDir = null
+    def androidSourcesName = 'React Native sources'
 
-  if (defaultDir.exists()) {
-    maven {
-      url defaultDir.toString()
-      name androidSourcesName
+    if (rootProject.ext.has('reactNativeAndroidRoot')) {
```
Comment (cortinico) [inline packages/create-react-native-library/templates/native-common/android/build.gradle:159]: What is `reactNativeAndroidRoot`? What is it used for?

**Review by cortinico — COMMENTED** [review]
In general, it looks right to me. Do you happen to have also a CMake file?

File: packages/create-react-native-library/templates/native-common/android/build.gradle (hunk 3 above)
Comment (satya164) [reply packages/create-react-native-library/templates/native-common/android/build.gradle:159]: @cortinico it would've been useful for cases such as monorepo or non-standard project structure when we couldn't locate the `react-native` folder, but actually, we don't need this here anymore since the `REACT_NATIVE_NODE_MODULES_DIR` (that I adapted from reanimated) config does the same thing. I'll remove this one.

File: packages/create-react-native-library/templates/native-common/android/build.gradle (hunk 2 above)
Comment (satya164) [reply packages/create-react-native-library/templates/native-common/android/build.gradle:150]: @cortinico at least for code generated by bob, the absence of this folder wouldn't cause an issue. previously it was happy as long as it found the `react-native` folder in `node_modules` - though the original code was from `react-native-webview` repo so I don't have the context of why.

File: packages/create-react-native-library/templates/native-common/android/build.gradle (hunk 1 above)
Comment (satya164) [reply packages/create-react-native-library/templates/native-common/android/build.gradle:186]: will remove after rc3 is released since currently this is needed to avoid build error.

**Comment by satya164** [comment]
> Do you happen to have also a CMake file?

We have this file if the library is using cpp code: https://github.com/callstack/react-native-builder-bob/blob/main/packages/create-react-native-library/templates/cpp-library/android/CMakeLists.txt

File: packages/create-react-native-library/templates/native-common/android/build.gradle (hunk 3 above)
Comment (cortinico) [reply packages/create-react-native-library/templates/native-common/android/build.gradle]: Got it thank you 👍 I suspected that this was the case.

File: packages/create-react-native-library/templates/native-common/android/build.gradle [hunk 4]
```diff
@@ -19,41 +19,81 @@ buildscript {
...
+  try {
+    // Resolve React Native location with Node
+    // This will make sure that we get installation location correctly in monorepos
+    def reactNativePackageJsonPathStdout = new ByteArrayOutputStream()
+
+    exec {
+      commandLine("node", "-p", "require.resolve('react-native/package.json')")
+      ignoreExitValue true
+      standardOutput = reactNativePackageJsonPathStdout
+    }
```
Comment (cortinico) [inline packages/create-react-native-library/templates/native-common/android/build.gradle:41]: Q: Why is this necessary now? How was this working on 0.70?
I would refrain from doing `exec{}` here as they're slowing down every build, for everyone and they get called for every library, so this gets exponentially worse.

**Comment by satya164** [comment]
@cortinico no worries, i'll wait for any feedback before releasing this

File: packages/create-react-native-library/templates/native-common/android/build.gradle (hunk 4 above)
Comment (cortinico) [reply packages/create-react-native-library/templates/native-common/android/build.gradle:41]: > we check the react native version to specify the correct dependency

How were you checking the rn version before .71?

Comment (satya164) [reply packages/create-react-native-library/templates/native-common/android/build.gradle:41]: @cortinico previously we didn't need to do that. we check it now since some of the code is different based on the version. Here is a snippet from generated gradle config:

```gradle

def isNewArchitectureEnabled() {
  return rootProject.hasProperty("newArchEnabled") && rootProject.getProperty("newArchEnabled") == "true"
}

def resolveReactNativeDirectory() {
  // ...
}

apply plugin: 'com.android.library'
apply plugin: 'kotlin-android'

if (isNewArchitectureEnabled()) {
  apply plugin: 'com.facebook.react'
}

def getExtOrDefault(name) {
  return rootProject.ext.has(name) ? rootProject.ext.get(name) : project.properties['ReactNativeTest_' + name]
}

def getExtOrIntegerDefault(name) {
  return rootProject.ext.has(name) ? rootProject.ext.get(name) : (project.properties['ReactNativeTest_' + name]).toInteger()
}

def reactNativeRootDir = resolveReactNativeDirectory()
def reactProperties = new Properties()
file("$reactNativeRootDir/ReactAndroid/gradle.properties").withInputStream { reactProperties.load(it) }

def REACT_NATIVE_VERSION = reactProperties.getProperty("VERSION_NAME")
def REACT_NATIVE_MAJOR_VERSION = REACT_NATIVE_VERSION.split("\\.")[0].toInteger()
def REACT_NATIVE_MINOR_VERSION = REACT_NATIVE_VERSION.split("\\.")[1].toInteger()

android {
  compileSdkVersion getExtOrIntegerDefault('compileSdkVersion')

  defaultConfig {
    minSdkVersion getExtOrIntegerDefault('minSdkVersion')
    targetSdkVersion getExtOrIntegerDefault('targetSdkVersion')
    buildConfigField "boolean", "IS_NEW_ARCHITECTURE_ENABLED", isNewArchitectureEnabled().toString()
  }
  buildTypes {
    release {
      minifyEnabled false
    }
  }

  lintOptions {
    disable 'GradleCompatible'
  }

  compileOptions {
    sourceCompatibility JavaVersion.VERSION_1_8
    targetCompatibility JavaVersion.VERSION_1_8
  }

  sourceSets {
    main {
      if (isNewArchitectureEnabled()) {
          java.srcDirs += [
            // This is needed to build Kotlin project with NewArch enabled
            "${project.buildDir}/generated/source/codegen/java"
          ]
      }
    }
  }
}

repositories {
  mavenCentral()
  google()

  if (REACT_NATIVE_MAJOR_VERSION == 0 && REACT_NATIVE_MINOR_VERSION < 71) {
    def androidSourcesDir = null
    def androidSourcesName = 'React Native sources'

    if (rootProject.ext.has('reactNativeAndroidRoot')) {
      androidSourcesDir = rootProject.ext.get('reactNativeAndroidRoot')
    } else {
      androidSourcesDir = new File(reactNativeRootDir, '/android')
    }

    if (androidSourcesDir.exists()) {
      maven {
        url androidSourcesDir.toString()
        name androidSourcesName
      }

      logger.info("${project.name}:reactNativeAndroidRoot ${defaultDir.canonicalPath}")
    }
  }
}

def kotlin_version = getExtOrDefault('kotlinVersion')

dependencies {
  if (REACT_NATIVE_MAJOR_VERSION == 0 && REACT_NATIVE_MINOR_VERSION < 71) {
    //noinspection GradleDynamicVersion
    implementation 'com.facebook.react:react-native:+'
    // From node_modules
  } else {
    implementation 'com.facebook.react:react-android:' + REACT_NATIVE_VERSION
  }

  implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
}

if (isNewArchitectureEnabled()) {
  react {
    jsRootDir = file("../src/")
    libraryName = "ReactNativeTestView"
    codegenJavaPackageName = "com.bob.reactnativetest"
  }
}
```

Comment (cortinico) [reply packages/create-react-native-library/templates/native-common/android/build.gradle:41]: Thanks for clarifying. Yup that's a problem and we should look into it. Ideally the React Native Gradle Plugin should expose this information and you should be able to just consume it.

Comment (satya164) [reply packages/create-react-native-library/templates/native-common/android/build.gradle:41]: @cortinico Gotcha! I'll keep an eye out and help with testing it

Comment (cortinico) [reply packages/create-react-native-library/templates/native-common/android/build.gradle:41]: So back on this, I think this whole logic can be cleaned to don't need `REACT_NATIVE_MINOR_VERSION` at all.

1. The `implementation` dependency can be just `implementation 'com.facebook.react:react-native:+'`. On 71+ the Gradle Plugin will take care of handling this.
2. The repositories block should be just "safe" in the sense that if the `/android` folder can't be found => No extra `maven{}` block should be called (which is already the case).

Comment (cortinico) [reply packages/create-react-native-library/templates/native-common/android/build.gradle:41]: I'll keep this in mind for the future and for .72 we might be able to expose `reactNativeDir` and `reactNativeVersion` to libraries via the Gradle Plugin
